tensorflow/core/kernels/non_max_suppression_op.cc

```diff
@@ -284,14 +284,176 @@ void DoNonMaxSuppressionOp(OpKernelContext* context, const Tensor& scores,
+                          bool clip_boxes, int per_batch_size) {
+  auto rc_cmp = [](const ResultCandidate rc_i, const ResultCandidate rc_j) {
+    return rc_i.score > rc_j.score;
+  };
+  std::sort(result_candidate_vec.begin(), result_candidate_vec.end(), rc_cmp);
+
+  int max_detections = 0;
+  int result_candidate_size =
+      std::count_if(result_candidate_vec.begin(), result_candidate_vec.end(),
+                    [](ResultCandidate rc) { return rc.box_index > -1; });
+  // If pad_per_class is false, we always pad to max_total_size
+  if (!pad_per_class) {
+    max_detections = std::min(result_candidate_size, total_size_per_batch);
+  } else {
+    max_detections = std::min(per_batch_size, result_candidate_size);
+  }
+
+  final_valid_detections[batch_idx] = max_detections;
+
+  int curr_total_size = max_detections;
+  int result_idx = 0;
+  // Pick the top max_detections values
+  while (curr_total_size > 0 && result_idx < result_candidate_vec.size()) {
+    ResultCandidate next_candidate = result_candidate_vec[result_idx++];
+    // Add to final output vectors
+    if (clip_boxes) {
+      const float box_min = 0.0;
+      const float box_max = 1.0;
+      nmsed_boxes.push_back(
+          std::max(std::min(next_candidate.box_coord[0], box_max), box_min));
+      nmsed_boxes.push_back(
+          std::max(std::min(next_candidate.box_coord[1], box_max), box_min));
+      nmsed_boxes.push_back(
+          std::max(std::min(next_candidate.box_coord[2], box_max), box_min));
+      nmsed_boxes.push_back(
+          std::max(std::min(next_candidate.box_coord[3], box_max), box_min));
+    } else {
+      nmsed_boxes.push_back(next_candidate.box_coord[0]);
+      nmsed_boxes.push_back(next_candidate.box_coord[1]);
+      nmsed_boxes.push_back(next_candidate.box_coord[2]);
+      nmsed_boxes.push_back(next_candidate.box_coord[3]);
+    }
+    nmsed_scores.push_back(next_candidate.score);
+    nmsed_classes.push_back(next_candidate.class_idx);
+    curr_total_size--;
+  }
+
+  nmsed_boxes.resize(per_batch_size * 4, 0);
+  nmsed_scores.resize(per_batch_size, 0);
+  nmsed_classes.resize(per_batch_size, 0);
+}
+
 void BatchedNonMaxSuppressionOp(
     OpKernelContext* context, const Tensor& inp_boxes, const Tensor& inp_scores,
     int num_boxes, const int max_size_per_class, const int total_size_per_batch,
     const float score_threshold, const float iou_threshold,
     bool pad_per_class = false, bool clip_boxes = true) {
-  int q = inp_boxes.dim_size(2);
-  int num_classes = inp_scores.dim_size(2);
   const int num_batches = inp_boxes.dim_size(0);
+  int num_classes = inp_scores.dim_size(2);
+  int q = inp_boxes.dim_size(2);
+
+  const float* scores_data =
+      const_cast<float*>(inp_scores.flat<float>().data());
+  const float* boxes_data = const_cast<float*>(inp_boxes.flat<float>().data());
+
+  int boxes_per_batch = num_boxes * q * 4;
+  int scores_per_batch = num_boxes * num_classes;
```

Comment:
These gave unused variables errors. Please capture them in `shard_nms` and delete lines 468-469. Thank you very much!

Reply:
@penpornk Done, thanks.